javatests/arcs/core/data/expression/TypeCheckerTest.kt

```diff
@@ -0,0 +1,300 @@
+    checkAssignableTo(1.toShort().asExpr(), LongType)
+    checkAssignableTo(1.toShort().asExpr(), FloatType)
+    checkAssignableTo(1.toShort().asExpr(), DoubleType)
+    checkAssignableTo(1.toShort().asExpr(), NumberType)
+
+    checkAssignableTo(1.toByte().asExpr(), ByteType)
+    checkAssignableTo(1.toByte().asExpr(), ShortType)
+    checkAssignableTo(1.toByte().asExpr(), IntType)
+    checkAssignableTo(1.toByte().asExpr(), BigIntType)
+    checkAssignableTo(1.toByte().asExpr(), LongType)
+    checkAssignableTo(1.toByte().asExpr(), FloatType)
+    checkAssignableTo(1.toByte().asExpr(), DoubleType)
+    checkAssignableTo(1.toByte().asExpr(), NumberType)
+    checkAssignableTo("foo".asExpr(), TextType)
+    checkAssignableTo(true.asExpr(), BooleanType)
+  }
+
+  @Test
+  fun check_unary_ops() {
+    checkAssignableTo(-(1.asExpr()), IntType)
+    checkAssignableTo(-(1.0.asExpr()), DoubleType)
+    checkAssignableTo(!(true.asExpr()), BooleanType)
+  }
+
+  @Test
+  fun check_binary_ops() {
+    checkAssignableTo(1.asExpr() + 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() + 2.asExpr(), DoubleType)
+
+    checkAssignableTo(1.asExpr() - 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() - 2.asExpr(), DoubleType)
+
+    checkAssignableTo(1.asExpr() * 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() * 2.asExpr(), DoubleType)
+
+    checkAssignableTo(1.asExpr() / 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() / 2.asExpr(), DoubleType)
+
+    checkAssignableTo(1.asExpr() eq 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() neq 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() lt 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() lte 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() gt 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() gte 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() eq 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() neq 2.asExpr(), BooleanType)
+
+    checkAssignableTo(true.asExpr() and true.asExpr(), BooleanType)
+    checkAssignableTo(true.asExpr() or true.asExpr(), BooleanType)
+
+    checkAssignableTo(1.asExpr() ifNull "hello".asExpr(), UnionType(NumberType, TextType))
+  }
+
+  @Test
+  fun check_field_ops() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("x", IntType)
+      set("y", DoubleType)
+      set("z", ScopeType(emptyScope.builder().set("a", TextType).build()))
+    }.build()
+    checkAssignableTo(num("x"), IntType, scope)
+    checkAssignableTo(num("y"), DoubleType, scope)
+    checkAssignableTo(scope("z").get<InferredType>("a"), TextType, scope)
+  }
+
+  @Test
+  fun check_query_ops() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("x", IntType)
+      set("y", DoubleType)
+    }.build()
+    checkAssignableTo(query<Number>("x"), IntType, scope, scope)
```

Comment:
put emptyScope as the ctx ?

i.e.
`checkAssignableTo(query<Number>("x"), IntType, emptyScope, scope)`